caravan.lua

```diff
@@ -1,31 +1,311 @@
+end
+
+function collapseAllTypes()
+   collapseTypes(df.global.game.main_interface.trade.current_type_a_expanded[0], 0)
+   collapseTypes(df.global.game.main_interface.trade.current_type_a_expanded[1], 1)
+   -- reset scroll to top when collapsing types
+   df.global.game.main_interface.trade.scroll_position_item[0] = 0
+   df.global.game.main_interface.trade.scroll_position_item[1] = 0
+end
+
+function collapseContainers(item_list, list_index)
+    local num_items_collapsed = 0
+    for k, goodflag in ipairs(item_list) do
+        if goodflag ~= GOODFLAG.CONTAINED_UNSELECTED and goodflag ~= GOODFLAG.CONTAINED_SELECTED then
+            local next_item_index = k + 1
+            if next_item_index > #item_list - 1 then
+                goto skip
+            end
+
+            local next_item = item_list[next_item_index]
+            local this_item_is_container = df.item_binst:is_instance(df.global.game.main_interface.trade.good[list_index][k])
+            local collapsed_this_container = false
+            if this_item_is_container then
+                if goodflag == GOODFLAG.UNCONTAINED_SELECTED then
+                    item_list[k] = GOODFLAG.CONTAINER_COLLAPSED_SELECTED
+                    collapsed_this_container = true
+                elseif goodflag == GOODFLAG.UNCONTAINED_UNSELECTED then
+                    item_list[k] = GOODFLAG.CONTAINER_COLLAPSED_UNSELECTED
+                    collapsed_this_container = true
+                end
+
+                if collapsed_this_container then
+                    num_items_collapsed = num_items_collapsed + #dfhack.items.getContainedItems(df.global.game.main_interface.trade.good[list_index][k])
+                end
+            end
+
+            ::skip::
+        end
+    end
+
+    if num_items_collapsed > 0 then
+        df.global.game.main_interface.trade.i_height[list_index] = df.global.game.main_interface.trade.i_height[list_index] - num_items_collapsed * 3
+    end
+end
+
+function collapseAllContainers()
+    collapseContainers(df.global.game.main_interface.trade.goodflag[0], 0)
+    collapseContainers(df.global.game.main_interface.trade.goodflag[1], 1)
+end
+
+function collapseEverything()
+    collapseAllContainers()
+    collapseAllTypes()
+end
+
+function copyGoodflagState()
+    trader_selected_state = copyall(df.global.game.main_interface.trade.goodflag[0])
+    broker_selected_state = copyall(df.global.game.main_interface.trade.goodflag[1])
+end
+
+function CaravanOverlay:onInput(keys)
+    if not is_trade_overlay_active() then return false end
+    if keys._MOUSE_L then
+        if dfhack.internal.getModifiers().shift then
+            handle_shift_click_on_render = true
+        elseif dfhack.internal.getModifiers().ctrl then
+            handle_ctrl_click_on_render = true
+        end
+
+        if handle_ctrl_click_on_render or handle_shift_click_on_render then
+            copyGoodflagState()
+        end
+    end
+    CaravanOverlay.super.onInput(self, keys)
```

Comment:
it might be worthwhile to leave a comment here explaining that you are intentionally not returning a value from this method so everything is passed through to the underlying viewscreen. either that or explicitly return false.